### DISCUSSION TITLE
Use Machine Learning for document separation

### DISCUSSION THREAD

**Discussion by johnzielke** [paperless-ngx/paperless-ngx#2708] [Feature Requests]
I was wondering if anyone ever looked into using Machine Learning for automatically recognizing individual documents in a batch of pages that were all scanned together.
I saw that one can use the PATCH-T page separators, but these require manually adding barcode pages to the stack between documents before scanning them in the ADF.
Using ML this could be automated. Given we had such a model (although training it would require a suitable dataset already split into documents), would you think it would be helpful to integrate such a model into paperless-ngx?
My main questions would be:
1. Would it even be feasible to add such a model as it would require more powerful hardware to run such models
2. Given experience of actually using the system, what of the following approaches works best and how time intensive are they: Adding the barcode pages, manually splitting after the scan using pdfsam or similar.

This would help me better understand if an ML model would be helpful and I could maybe build and integrate it into paperless-ngx.

**Comment by tomihbk** [comment]
Just my two cents :

What defines a stack of papers to be separated ? There must be a certain type of pattern to know this from POV of an ML. With the multiple types of documents, different correspondents, tags, many more. All documents are dynamic, I don't think it would be easy to implement this. Even if you have to train the machine, it would need unholy amount of data.

**Comment by jonaswinkler** [comment]
Adding my two cents as well, and I do have some background in ML.

Typical documents scanned into paperless are letters, invoices, receipts, contracts, and so on. Those documents usually have some sort of header on the first page, block with addresses, subject, name of some vendor, etc. Lets say we have a set of documents that is already properly separated. Use those documents in the following way:

Train a model that, as input, takes the text of a single page. The output is a confidence value between 0 and 1, where 1 means that it is the first page in the document. Training data consists of all individual pages, and the output during training is 1 for the first page, 0 otherwise. For any new document, run the model on each page. If it yields 1, split at this page.

This doesn't necessarily need an unholy amount of data. If we assume that one user typically scans the same kind of documents, then the usual amount of documents in a paperless instance (lets say 300+ as an arbitrary number) with enough variance in document types and correspondents and a large percentage of multi-page documents should be enough to get good results.

A PoC shouldn't be that hard to make with python, pikepdf, and sklearn. Integrating into paperless is another topic though, and largely depends on how accurate this is. I'd say implementing this is unlikely, since correcting errors in page separation is way more tedious than correcting a few tags or correspondents.

**Reply by jonaswinkler** [reply]
(However, I want to start picking up work on this split and merge feature again, and I wouldn't reject the possibility of a "magic split" button in the interface at some point in the future.)

**Comment by johnzielke** [comment]
My background is in ML, so I am trying to get some "domain" knowledge and insights into what the benefits of the model would even be and what failure modes would be acceptable and which not.
The alternative to the Machine Learning approach would be to go through the scan manually and mark each page split (with the right tool, that's a single button press per page split). So to be able to beat that (and assuming you have a lot of smaller (e.g. around 4 page) documents) the system needs to be good enough to work without any review of the documents, or do you think there is a big benefit to having all the page splits preselected, but still manually reviewing them?

I think one question regarding accuracy is: Would you expect any other person that knows nothing about your documents to be able to split a stack of scanned documents correctly? If so, I think an ML model would be able to do it as well with good accuracy.
The question is on what data this model would be trained and how it would work.
Some options are:
- A pretrained model using the actual image of the page (optionally together with the OCRed text). This looking at the page would enable it to take logos and other layout information into account. This would be tough to run on really weak hardware, but I guess having optional features that you can only enable with good hardware are not impossible. More importantly, training this model would require a corpus of documents to train it, and then provide the model weights to users. The data would not need to be public, but it would be preferable.
- Having a image-based model trained for each user on their existing data. If the local data pool is big enough, this could be enough to not require a central dataset for pretraining and would allow the model to adjust to the user's data. But training such a model is very computation-intensive and would therefore be restricted to people with decent hardware and would be a nightmare to support with all different hardware configurations, therefore I think this is not an option.
- Use a simple purely text-based model using some embedding or similarity metric. This would be computationally and implementation-wise much easier, but will probably be the least accurate option.

**Reply by sense4t** [reply]
my 2 cents:
this should definitely become an optional setting. 
What makes a document a document?
I have lots of scans of letters with attachments, like tax statements, which should remain a single stack and not become separated into single pages. I'm not sure whether ML could distinguish my intention to keep things together ...

**Comment by HolgerM1** [comment]
Hi,
I just started using Paperless. I'm coming from another DMS (EcoDMS). In EcoDMS you could use a given separator page [https://www.ecodms.de/index.php/de/download/ecodms-archiv/ecodms-burns/ecodms-build-22-08](url). I liked using this because usually I wait until I have a bunch of documents that I drop in my scanner. Maybe that would be an option?!

**Reply by ph1248** [reply]
That's already implemented. For a start, see the configuration options in the [docs](https://docs.paperless-ngx.com/configuration/#polling) for 
`PAPERLESS_CONSUMER_ENABLE_BARCODES` and `PAPERLESS_CONSUMER_BARCODE_STRING`

However, at the moment it does not always work reliably, see e.g. #2678 . I am currently experimenting with it and will post there if I have found out anything new.

**Comment by github-actions** [comment]
This discussion has been automatically closed due to lack of community interest.

**Comment by github-actions** [comment]
This discussion has been automatically locked since there has not been any recent activity after it was closed. Please open a new discussion for related concerns. See our [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/dev/CONTRIBUTING.md#automatic-repository-maintenance) for more details.
